### ISSUE TITLE
Builds are kind of not using Hyper Threading

### ISSUE THREAD

**Issue by varosi** [commercialhaskell/stack#4391]
### Steps to reproduce

For example:

1. stack install hoogle on a clean configuration
2. Most of the time when all the cores are utilised it will be without any usage of Hyper Threading.

Include any `.yaml` configuration if relevant.

### Expected

To use all available logical cores, not just physical ones.

### Actual

Currently most build processes are fully utilising only single logical thread per core.
Tested on iMac and MacBook Pro with latest OSX Mojave.
See this video: https://drive.google.com/file/d/1YRPBURxrj6pr5AeTepUm3iCXwgvM4Hj7/view?usp=sharing
On the top you could see in green/red the logical cores of the system. Most of the time they are utilizing to only half of all (6 cores with HT = 12 logical cores). If more cores are used we would see that more than 6 are utilised. From time to time there are used all 12, but either they are less than 6, or they are 12. So I think that GHC build processes run on 6 cores and from time to time we could observe two GHC processes with 6 cores. It's strange.

Do someone else observe that on Mac? I'm even not sure it's from Stack. It could be Cabal or GHC build. But I'm trying to start from somewhere.

### Stack version

```
$ stack --version
Version 1.9.1, Git revision f9d0042c141660e1d38f797e1d426be4a99b2a3c (6168 commits) x86_64 hpack-0.31.0
```

### Method of installation

* Official binary

**Comment by varosi-chaosgroup** [comment]
There are no such problems on Windows 10 x64:
![image](https://user-images.githubusercontent.com/31626390/48188336-3919ad00-e346-11e8-8656-f341ee615285.png)

**Comment by mattaudesse** [comment]
@varosi that does seem strange, but I think you're probably right about it being upstream of `stack` and having to due with `cabal` and/or `ghc`.

Does running with `stack install --jobs 12 hoogle` match the desired behavior you're expecting? I believe `stack` uses maximal concurrency by default, so I would've thought it would supply the `--jobs 12` argument for you transparently. Perhaps the mechanism for determining how many logical cores are available is buggy(?)

On my 2013 MacBook Air (a dual-core Intel i5) running Mojave I saw all four logical cores being pretty much saturated without using any additional configuration.

**Comment by snoyberg** [comment]
Are there any next steps here? Has this been identified as a GHC issue?

**Comment by varosi** [comment]
I'm not tried it in pure GHC. May be I could get some cabal project to test on. From a long time I use Stack tool for building and I even almost forget how to use Cabal :-D

**Comment by varosi** [comment]
Cabal is utilising CPU much more tighter than Stack as of today.
See Cabal:
![Screen Shot 2019-04-16 at 10 25 13](https://user-images.githubusercontent.com/6051718/56190120-3d06d500-6032-11e9-9614-74938bbd98e9.png)

See Stack:
![Screen Shot 2019-04-16 at 10 25 50](https://user-images.githubusercontent.com/6051718/56190125-4001c580-6032-11e9-9e20-90f449a1833e.png)

Stack seems to run multiple processes, but each one separately doesn't use Hyper Threading. So it looks like the OS tries to distribute two Non-HT processes to use more of the CPU power.

**Comment by varosi** [comment]
Possibly CPU logical processors count detection on macOS is not done right or something in this manner.

**Comment by snoyberg** [comment]
My only guess about what's happening here would be that Cabal is using component-based builds, while Stack isn't yet. Component based builds may allow better parallelism. We're planning on including that soon, so it would be interesting to retest this once it lands.

**Comment by varosi-chaosgroup** [comment]
Is there an issue about it, so we could link it here?

**Comment by snoyberg** [comment]
Turns out we only tangentially addressed it. New issue opened at #4745

**Comment by varosi** [comment]
Great! Thanks!

**Comment by varosi** [comment]
I have tried to build latest Stack with Stack 2.0.0 and it still totally miss Hyper Threading for some reason during dependencies building:
<img width="170" alt="Screenshot 2019-05-23 at 2 16 41" src="https://user-images.githubusercontent.com/6051718/58215162-2aeb1700-7d01-11e9-88bc-bee4fc4d915b.png">

As it seen in the image - even on full build power it uses every second logical core without touching half of available ones.

**Comment by varosi** [comment]
I confirm this issue is still present in 2.1.0.3 (Release Candidate)

**Comment by borsboom** [comment]
I just tried it on my quad-core Mac and I do *not* observe this problem (using stack-2.1.0.3).  All eight hyper-threads are utilized.

<img width="568" alt="Screen Shot 2019-06-04 at 5 51 09 AM" src="https://user-images.githubusercontent.com/2164261/58880483-d3e42b00-868c-11e9-9718-9f6a6e5acb3d.png">

<img width="245" alt="Screen Shot 2019-06-04 at 5 52 27 AM" src="https://user-images.githubusercontent.com/2164261/58880542-fd04bb80-868c-11e9-95d4-0884a37a878d.png">

**Comment by varosi-chaosgroup** [comment]
Are they utilised most of the time when there is a pressure on the CPU? 
I observe that when there is a pressure either you get all logical cores utilised, or exactly half of the cores and each on 100% (nearby cores are not, at that time). This makes me think that Stack is spawning processes that doesn't use HT on macOS, but if it spawn two processes and each load half of logical cores (separately those processes doesn't detect HT, so they see half of logical cores) - then the OS is placing each process with its half logical cores and total result is the observation.

**Comment by varosi-chaosgroup** [comment]
This behaviour is not observed on Windows.

**Comment by varosi-chaosgroup** [comment]
May be component based builds will fix that issue. 2.x is more performant and even now it's better than before.
